### PR TITLE
[WIP] Add `Loop.stop()`

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Add `Loop.stop()` to stop a loop and replace `trainer.should_stop`

Necessary for #8578 

### Does your PR introduce any breaking changes? If yes, please list them.

TODO

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: pytorch_lightning/loops/base.py [hunk 1]
```diff
@@ -103,18 +104,29 @@ def run(self, *args: Any, **kwargs: Any) -> Optional[Any]:
...
-                self.on_advance_end()
+                if not self.should_stop:
+                    self.advance(*args, **kwargs)
+                if not self.should_stop:
+                    self.on_advance_end()
                 self.restarting = False
             except StopIteration:
                 break
 
         output = self.on_run_end()
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/loops/base.py:118]: We might decide to stop on the latest batch on on_advance_end.

File: pytorch_lightning/loops/fit_loop.py [hunk 2]
```diff
@@ -124,31 +124,23 @@ def _results(self) -> ResultCollection:
...
-                    "Trainer was signaled to stop but required minimum epochs"
-                    f" ({self.min_epochs}) or minimum steps ({self.min_steps}) has"
-                    " not been met. Training will continue..."
-                )
-        self.trainer.should_stop = should_stop
-
-        return stop_steps or should_stop or stop_epochs
+        return stop_steps or stop_epochs
+
+    def stop(self) -> None:
```
Comment (tchaton) [inline pytorch_lightning/loops/fit_loop.py:133]: Really neat !

**Comment by tchaton** [comment]
@carmocca really like the design :)

**Comment by carmocca** [comment]
Note: this feature would have avoided the need to add a `TunerExitException` in #11089 to stop the loops. Additionally being public loops API.

I'm not continuing this work atm as the future of the loops is unclear at this moment.

**Pull request closed without merging**